import add_ep_kind from './extensions.jsy'

add_ep_kind @:
  api_bind_rpc
  api(api) :: return this.api_parallel(api)
  api_parallel(api) ::
    return this.endpoint @ function (ep, hub) ::
      const rpc = api_bind_rpc(api, ep, hub)
      return @{} rpc,
        async on_msg({msg, sender}) ::
          await rpc.invoke @ sender, msg.op,
            api_fn => api_fn(msg.kw, msg.ctx)

  api_inorder(api) ::
    return this.endpoint @ function (ep, hub) ::
      const rpc = api_bind_rpc(api, ep, hub)
      return @{} rpc,
        async on_msg({msg, sender}) ::
          await rpc.invoke_gated @ sender, msg.op,
            api_fn => api_fn(msg.kw, msg.ctx)


function api_bind_rpc(api, ep, hub) ::
  const pfx = api.op_prefix || 'rpc_'
  const lookup_op = api.op_lookup
    ? op => api.op_lookup(pfx + op, ep, hub)
    : 'function' === typeof api
    ? op => api(pfx + op, ep, hub)
    : op => ::
        const fn = api[pfx + op]
        return fn ? fn.bind(api) : fn

  return Object.create @ rpc_api, @{}
    lookup_op: @{} value: lookup_op
    err_from: @{} value: ep.ep_self()


const rpc_api = @:
  async invoke(sender, op, cb) ::
    const api_fn = await this.resolve_op @ sender, op
    if undefined === api_fn :: return

    const res = this.answer @ sender, api_fn, cb
    return await res

  async invoke_gated(sender, op, cb) ::
    const api_fn = await this.resolve_op @ sender, op
    if undefined === api_fn :: return

    const res = Promise.resolve(this.gate)
      .then @ () => this.answer @ sender, api_fn, cb
    this.gate = res.then(noop, noop)
    return await res

  async resolve_op(sender, op) ::
    if 'string' !== typeof op ::
      await sender.send @: op, err_from: this.err_from
        error: @{} message: 'Invalid operation', code: 400
      return

    try ::
      const api_fn = await this.lookup_op(op)
      if ! api_fn ::
        await sender.send @: op, err_from: this.err_from
          error: @{} message: 'Unknown operation', code: 404
      return api_fn
    catch err ::
      await sender.send @: op, err_from: this.err_from
        error: @{} message: `Invalid operation: ${err.message}`, code: 500

  async answer(sender, api_fn, cb) ::
    try ::
      var answer = cb ? await cb(api_fn) : await api_fn()
    catch err ::
      await sender.send @: err_from: this.err_from, error: err
      return false

    if sender.replyExpected ::
      await sender.send @: answer
    return true


function noop() {}

